### PR TITLE
[Ready to Review] Feature/tags should be an array

### PR DESCRIPTION
**Pull request by pattisdr**
# Purpose

Issue https://openscience.atlassian.net/browse/OSF-4602

When creating/updating a node, a dictionary is allowed as input.  We need to enforce that the type is an array.

PUTting {"tags": {"foo":"bar", "baz":"quux"}} into /v2/nodes/$node_id results in the keys of the object ("foo", "baz") being used as the tags and the values being ignored. 

# Changes
Creates new ListField which overrides to_internal_value, enforcing that the data type is a list. 

# For QA
Try to create a node, fill in all required fields.  For tags, put a dictionary.  
screen shot 2015-10-09 at 2 22 53 pm
Expected response:
screen shot 2015-10-09 at 2 23 36 pm

And if tags are the correct format, a list, node will be created successfully with tags.
screen shot 2015-10-09 at 2 25 21 pm

### CODE REVIEW

File: api/base/serializers.py [hunk 1]
```diff
@@ -79,6 +79,14 @@ def to_internal_value(self, data):
         return super(TypeField, self).to_internal_value(data)
 
 
+class JSONAPIListField(ser.ListField):
+    def to_internal_value(self, data):
+        if not isinstance(data, list):
+            self.fail('not_a_list', input_type=type(data).__name__)
+
+        return super(JSONAPIListField, self).to_internal_value(data)
```
Comment (abought) [inline api/base/serializers.py]: Slightly more specific case of the error check in the parent class (`ser.ListField.to_internal_value`); makes sense. :octopus: :palm_tree:

File: tests/api_tests/nodes/test_views.py [hunk 2]
```diff
@@ -2603,6 +2603,26 @@ def test_tags_add_and_remove_properly(self):
...
+        res = self.app.post_json_api(url, payload, auth=self.user.auth, expect_errors=True)
+        assert_equal(res.status_code, 400)
+        assert_equal(res.json['errors'][0]['detail'], 'Expected a list of items but got type "dict".')
+
+    def test_tags_patch_object_instead_of_list(self):
+        self.one_new_tag_json['data']['attributes']['tags'] = {'foo': 'bar'}
+        res = self.app.patch_json_api(self.public_url, self.one_new_tag_json, auth=self.user.auth, expect_errors=True)
+        assert_equal(res.status_code, 400)
+        assert_equal(res.json['errors'][0]['detail'], 'Expected a list of items but got type "dict".')
+
```
Comment (abought) [inline tests/api_tests/nodes/test_views.py]: I see POST and PATCH tests; is it safe to assume that nothing unusual will happen for PUT?

**Comment by abought** [comment]
CR1 complete; no changes requested.

**Comment by samanehsan** [comment]
Looks good to me @pattisdr! CR2 complete :fallen_leaf:

File: api/base/serializers.py (hunk 1 above)
Comment (sloria) [inline api/base/serializers.py]: This appears to be a bug in Django Rest Framework's ListField.

```python
    def to_internal_value(self, data):
        #...
        if isinstance(data, type('')) or not hasattr(data, '__iter__'):
                self.fail('not_a_list', input_type=type(data).__name__)
```

It is not properly checking if `data` is dict/mapping.

@pattisdr Would you like to report this issue and/or send a PR to DRF?

Comment (pattisdr) [inline api/base/serializers.py]: Yes, I will take care of this.

**Pull request merged**